pkg/distributor/distributor.go

```diff
@@ -1846,6 +1852,57 @@ func (d *Distributor) ActiveSeries(ctx context.Context, matchers []*labels.Match
 	return deduplicatedSeries, nil
 }
 
+type entry struct {
+	first      labels.Labels
+	collisions []labels.Labels
+}
+
+// activeSeriesResponse is a helper to merge/deduplicate ActiveSeries responses from ingesters.
+type activeSeriesResponse struct {
+	m                  sync.Mutex
+	series             map[uint64]entry
+	builder            labels.ScratchBuilder
+	lbls               labels.Labels
+	hashCollisionCount prometheus.Counter
+}
+
+func newActiveSeriesResponse(hashCollisionCount prometheus.Counter) *activeSeriesResponse {
+	return &activeSeriesResponse{
+		series:             map[uint64]entry{},
+		builder:            labels.NewScratchBuilder(40),
+		hashCollisionCount: hashCollisionCount,
+	}
+}
+
+func (r *activeSeriesResponse) add(series []*mimirpb.Metric) {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	for _, metric := range series {
+		mimirpb.FromLabelAdaptersOverwriteLabels(&r.builder, metric.Labels, &r.lbls)
+		lblHash := r.lbls.Hash()
+		if e, ok := r.series[lblHash]; !ok {
+			r.series[lblHash] = entry{first: r.lbls.Copy()}
+		} else if !labels.Equal(e.first, r.lbls) {
+			e.collisions = append(e.collisions, r.lbls.Copy())
+			r.series[lblHash] = e
+			r.hashCollisionCount.Inc()
+		}
```

Comment:
Oh yes, this is something I missed in my pseudo-code. We should also check whether `r.lbls` is one of the `e.collisions` 🙏 .

Can you add a test with collisions to make sure we have this path covered?

Reply:
Good point, done in https://github.com/grafana/mimir/pull/6717/commits/bedf6889d675642b5c85f5e3d46a976dc1fd24ea. I believe we don't have test data for a three-way collision, so the current test only covers the one-collision path.